src/common/providers/identity.ts

```diff
@@ -109,3 +375,598 @@ export function userRecordConstructor(wireData: Object): UserRecord {
+    throw new HttpsError(
+      'invalid-argument',
+      'Request has invalid header Content-Type.'
+    );
+  }
+
+  if (!req.body || !req.body.data || !req.body.data.jwt) {
+    throw new HttpsError('invalid-argument', 'Request has an invalid body.');
+  }
+}
+
+/** @internal */
+export function getPublicKey(
+  header: Record<string, any>,
+  publicKeys: Record<string, string>
+): string {
+  if (header.alg !== JWT_ALG) {
+    throw new HttpsError(
+      'invalid-argument',
+      `Provided JWT has incorrect algorithm. Expected ${JWT_ALG} but got ${header.alg}.`
+    );
+  }
+  if (!header.kid) {
+    throw new HttpsError('invalid-argument', 'JWT has no "kid" claim.');
+  }
+  if (!publicKeys.hasOwnProperty(header.kid)) {
+    throw new HttpsError(
+      'invalid-argument',
+      'Provided JWT has "kid" claim which does not correspond to a known public key. Most likely the JWT is expired.'
+    );
+  }
+
+  return publicKeys[header.kid];
+}
+
+/**
+ * @internal
+ * Checks for a well forms cloud functions url
+ */
+export function isAuthorizedCloudFunctionURL(
+  cloudFunctionUrl: string,
+  projectId: string
+): boolean {
+  // Region can be:
+  // us-central1, us-east1, asia-northeast1, europe-west1, asia-east1.
+  // Sample: https://europe-west1-fb-sa-upgraded.cloudfunctions.net/function-1
+  const gcf_directions = [
+    'central',
+    'east',
+    'west',
+    'south',
+    'southeast',
+    'northeast',
+    // Other possible directions that could be added.
+    'north',
+    'southwest',
+    'northwest',
+  ];
+  const re = new RegExp(
+    `^https://[^-]+-(${gcf_directions.join(
+      '|'
+    )})[0-9]+-${projectId}\.cloudfunctions\.net/`
+  );
+  const res = re.exec(cloudFunctionUrl) || [];
+  return res.length > 0;
+}
+
+/**
+ * @internal
+ * Checks for errors in a decoded jwt
+ */
+export function checkDecodedToken(
+  decodedJWT: DecodedJwt,
+  projectId: string
+): void {
+  if (!isAuthorizedCloudFunctionURL(decodedJWT.aud, projectId)) {
+    throw new HttpsError(
+      'invalid-argument',
+      'Provided JWT has incorrect "aud" (audience) claim.'
+    );
+  }
+  if (decodedJWT.iss !== `${JWT_ISSUER}${projectId}`) {
+    throw new HttpsError(
+      'invalid-argument',
+      `Provided JWT has incorrect "iss" (issuer) claim. Expected ` +
+        `"${JWT_ISSUER}${projectId}" but got "${decodedJWT.iss}".`
+    );
+  }
+  if (typeof decodedJWT.sub !== 'string' || decodedJWT.sub.length === 0) {
+    throw new HttpsError(
+      'invalid-argument',
+      'Provided JWT has no "sub" (subject) claim.'
+    );
+  }
+  if (decodedJWT.sub.length > 128) {
+    throw new HttpsError(
+      'invalid-argument',
+      'Provided JWT has "sub" (subject) claim longer than 128 characters.'
+    );
+  }
+}
+
+/**
+ * @internal
+ * Verifies the jwt using the 'jwt' library and decodes the token with the public keys
+ * Throws an error if the event types do not match
+ */
+function verifyAndDecodeJWT(
```

Comment:
OOC, how do you plan to loosen the verification requirements for the emulator in the future?

Reply:
I was planning on checking for an environment variable, but until that decision gets made, I added in a function `shouldVerifyJWT()` to determine if we need to verify the JWT. Right now, it will always return `true`.